### PR TITLE
Detect non-ambiguous dangling XPath operators

### LINKED ISSUE #860: Reject XLSForm expressions ending with non-ambiguous dangling operators
#### Software and hardware versions

pyxform v4.5.0, Python 3.11

#### Problem description

pyxform accepts expressions that end with an operator that requires a
right-hand operand when external validation is disabled. For example:

| type    | name               | relevant          |
|---------|--------------------|-------------------|
| text    | source_value       |                   |
| integer | dependent_question | ${source_value} = |

Running conversion with `validate=False` or `xls2xform --skip_validate`
successfully generates an XForm containing:

```xml
<bind
    nodeset="/data/dependent_question"
    relevant="/data/source_value ="
    type="int"
/>
```

The expression is incomplete because the operator has no right-hand value or
expression. The same problem applies to dangling arithmetic, boolean, and union
operators.

This postpones the error until `ODK_Validate.jar` flags the generated XForm as
invalid, without identifying the original XLSForm cell.

#### Steps to reproduce the problem

1. Create an XLSForm containing the survey rows shown above.
2. Run:

   ```bash
   xls2xform form.xlsx form.xml --skip_validate
   ```

   Alternatively, call `convert(..., validate=False)`.

3. Observe that conversion succeeds and the incomplete expression is emitted
   into the generated XForm.

#### Expected behavior

pyxform should raise a `PyXFormError` before XForm generation, regardless of
whether external validation is enabled.

The error should identify the original sheet, workbook row, and column, for
example:

```text
[row : 3] On the 'survey' sheet, the 'relevant' value is invalid.
An operator must be followed by a value or expression.
```

#### Other information

The targeted validation should:

- Detect comparison operators (`=`, `!=`, `<`, `>`, `<=`, `>=`), arithmetic
  operators (`+`, `-`, `div`, `mod`), boolean operators (`and`, `or`), and the
  union operator (`|`), including trailing whitespace.
- Allow terminal `*` because it can be a valid XPath wildcard, as in `/data/*`.
- Allow operator words used as XPath node names, as in `/data/and`.
- Check expression-bearing survey, settings, and entities columns.
- Ignore disabled survey rows.
- Allow quoted operator characters and static text defaults.
- Preserve original sheet and header names in error messages.
- Run when `validate=False` and with `--skip_validate`.

pyxform's existing expression lexer can identify the final token, with context
used to distinguish operator words from XPath node names.

### PR DESCRIPTION
**Pull request by ukanga**
Closes #860

#### Why is this the best possible solution? Were any other approaches considered?

Expressions ending with an operator that requires a right-hand operand are
incomplete, but pyxform currently emits them into the generated XForm when
external validation is disabled. The resulting error is only reported later by
ODK Validate or a form engine, without the original XLSForm sheet, row, and
column information.

This change adds targeted validation before XForm generation. It uses the
existing expression lexer to inspect the final non-whitespace token and raises a
`PyXFormError` for these dangling operators:

- Comparisons: `=`, `!=`, `<`, `>`, `<=`, `>=`
- Arithmetic: `+`, `-`, `div`, `mod`
- Boolean: `and`, `or`
- Union: `|`

Terminal `*` is intentionally excluded because it can be a valid XPath
wildcard, such as `/data/*`. The keyword operators use token context so XPath
node names such as `/data/and` remain valid.

The validation covers:

- Survey bind expressions: `relevant`, `constraint`, `calculation`, `required`,
  and `readonly`.
- `choice_filter`, repeat count, and dynamic defaults.
- The settings `instance_name` expression.
- Entity `entity_id`, `create_if`, `update_if`, and `label` expressions.
- Aliased and normalized headers while retaining original sheet and column
  names in error messages.

It runs when using both `convert(validate=False)` and
`xls2xform --skip_validate`.

The targeted lexer check addresses unambiguously dangling operators while
leaving broader XPath validation unchanged.

#### What are the regression risks?

The main risk is incorrectly rejecting literal values or XPath node names that
contain operator text. This is mitigated by checking lexer tokens and the
surrounding token context rather than using a string suffix check.

Tests verify that:

- Every included operator is detected with no trailing whitespace, spaces, and
  mixed trailing whitespace.
- Complete expressions and quoted operator characters remain accepted.
- Terminal `*` wildcards and operator-word node names remain accepted.
- Disabled survey rows are ignored.
- Static defaults ending in an operator remain accepted.
- Dynamic defaults and other recognized expression columns are checked.
- Original sheet, row, and aliased column names are reported.
- ODK Validate is not invoked after this internal validation fails.

There are no public API or type-signature changes. The intentional behavior
change is that previously emitted invalid expressions now raise a
`PyXFormError`, including when external validation is skipped.

#### Does this change require updates to documentation? If so, please file an issue [here](https://github.com/XLSForm/xlsform.github.io) and include the link below.

No documentation updates are required. This is an internal validation
improvement for already invalid expressions.

#### AI/LLM assistance

This PR was developed with assistance from a coding agent and LLM, including
implementation, tests, and supporting text. The submitted changes remain the
responsibility of the contributor.

#### Before submitting this PR, please make sure you have:

- [x] included test cases for core behavior and edge cases in `tests`
- [x] run `python -m unittest` and verified all tests pass
- [x] run `ruff format pyxform tests` and `ruff check pyxform tests` to lint code
- [x] verified that any code or assets from external sources are properly credited in comments

### CODE REVIEW

**Comment by lognaturel** [comment]
Thanks for filing this, @ukanga. I think it's worth doing and the tests look reasonable. My immediate reaction is that it feels like the validation should be possible to do with an existing pass over fields that can have expressions. @lindsay-stevens can you please take a look when you get a chance?

File: pyxform/parsing/expression.py [hunk 1]
```diff
@@ -117,6 +117,21 @@ def parse_expression(text: str) -> tuple[Token, ...]:
...
+    return (
+        next(
+            (
+                token.type
+                for token in reversed(parse_expression(text))
+                if token.type != "WHITESPACE"
+            ),
+            None,
+        )
+        == "OPS_COMP"
```
Comment (lognaturel) [inline pyxform/parsing/expression.py]: Why would dangling comparison operators be more likely than any other dangling operator?

Comment (ukanga) [reply pyxform/parsing/expression.py]: Probably not; I started with an example I encountered. Let me check whether I can identify more examples and whether it makes sense to address them in this PR.

Comment (ukanga) [reply pyxform/parsing/expression.py]: It seems we have ambiguous paths and non-ambiguous paths. For the non-ambiguous, the proposed changes seem to align ok.

The `*` operator is considered ambiguous and, as such, is not addressed in this change.

I also tested the expression `${q1} = and ${q1} > 0` which is not flagged at any point including with ODK_Validate.jar. Is this valid?
```XML
<bind
      nodeset="/data/q2"
      type="int"
      relevant=" /data/q1  = and  /data/q1  &gt; 0"
  />
```

Comment (ukanga) [reply pyxform/parsing/expression.py]: @lognaturel 

The check has been expanded to cover arithmetic (`+`, `-`, `div`, `mod`), boolean (`and`, `or`), and union (`|`) operators in addition to comparisons.  Terminal `*` is excluded because it can be a valid XPath wildcard such as `/data/*`.

**Comment by lognaturel** [comment]
I briefly discussed this with @lindsay-stevens and we'd like to reconsider it after some of his ongoing work lands. He's looking at the general theme of issues around ${} references (https://github.com/XLSForm/pyxform/issues?q=state%3Aopen%20label%3A%22%24%7B%7D%20references%22) and there will be changes to the code this touches. We'll be sure to use the test and implementation as reference once this gets prioritized. Thanks again for filing the issue.
